wicker/core/datasets.py

```diff
@@ -22,6 +23,95 @@
 FILE_LOCK_TIMEOUT_SECONDS = 300
 
 
+def get_file_size_s3_multiproc(buckets_keys: List[Tuple[str, str]]) -> int:
+    """Get file size of s3 files, most often column files.
+
+    This works on any list of buckets and keys but is generally only
+    used for column files as those are the majority of what is stored on
+    s3 for Wicker. Wicker also stores parquet files on s3 but those are limited
+    to one file per dataset and one schema file.
+
+    This splits your buckets_keys_list across multiple processes on your local host
+    where each process is then multi threaded further. This reduces the i/o wait by
+    parellelizing across all available procs and threads on a single machine.
+
+    Args:
+        buckets_keys: (List[Tuple[str, str]]): A list of buckets and keys for which
+        to fetch size in bytes on s3. Tuple index 0 is bucket and index 1 is key of the file.
+
+    Returns:
+        int size of file list in bytes.
+    """
+    buckets_keys_chunks = chunk_data_for_split(chunkable_data=buckets_keys, chunk_number=200)
+
+    logging.info("Grabbing file information from s3 heads")
+    pool = Pool(cpu_count() - 1)
+    return sum(list(tqdm.tqdm(pool.map(get_file_size_s3_threaded, buckets_keys_chunks))))
+
+
+def get_file_size_s3_threaded(buckets_keys_chunks_local: List[Tuple[str, str]]) -> int:
+    """Get file size of a list of s3 paths.
+
+    Args:
+        buckets_keys_chunks_local - The list of tuples denoting bucket and key of files on s3 to
+        parse. Generally column files but will work with any data.
+
+    Returns:
+        int: size of the set of files in bytes
+    """
+    local_chunks = chunk_data_for_split(chunkable_data=buckets_keys_chunks_local, chunk_number=200)
+    thread_pool = ThreadPool()
+
+    return sum(list(tqdm.tqdm(thread_pool.map(iterate_bucket_key_chunk_for_size, local_chunks))))  # type: ignore
+
+
+def chunk_data_for_split(chunkable_data: List[Any], chunk_number: int = 500) -> List[List[Any]]:
```

Comment:
nit: Do we need to write this function? Is it possible to use some standard library like `itertools.islice`?

Reply:
Standard library like islice creates a generator at it's core. Normally this would be fine but `map` in the multi process module converts generators to lists implicitly as it broadcasts out all data at once. This means that it would all eventually end up as a list anyway. The accepted answer of what package to use is what you suggest but it comes with a large inefficiency here and the same end result but it's hidden behind python standard functions.

You could make an argument for using imap over map and then the generator would be consumed one at a time but it is significantly slower on proc startup that map. IMO using the function written and map is more efficient than using islice and imap, @convexquad what do you think?

Reply:
@aalavian I think that seeing the function written out here correctly doesn't bother me, even if it is not strictly super-Pythonic. Since I can scan the code with my eyes and see that the behavior for the last chunk is correct (either when there is no last chunk or it is undersized), I'd just say keep it but let me know what you think.

Reply:
I don't follow the efficiency argument since at worst case islice would be converted to List which is what this function already does and is a one time operation that should not take much (do we know it take much?)

I do not have a strong opinion on this nit comment, but generally that it is better not to write custom functions for standard utilities.

Reply:
You're spending cycles on creating the generator when it is doing the same thing as creating a list ultimately. Generators are not super cheap in python, they have a lot of overhead and are not great in all situations, especially when you are iterating over all elements and storing them in memory.